File: SWEA/Unrated/1767. ［SW Test 샘플문제］ 프로세서 연결하기/［SW Test 샘플문제］ 프로세서 연결하기.py
# import sys
# sys.stdin = open('input.txt','r')

T = int(input())
check_core = []
max_core = 0
min_line = 0
mv = [ (1, 0), (-1, 0), (0, 1), (0, -1)]

#상수
CORE = 1
LINE = 2
EMPTY = 0
MAX_LINE = 10**9

#해당 위치 기준으로 (시작점제외) 해당 방향으로 LINE 설치가 가능한가?
def can_connet(r, c, dir):
    dr, dc = mv[dir]
    nr, nc = r + dr, c + dc
    while True:
        if nr < 0 or nr >= N or nc < 0 or nc >= N:
            break
        if lst[nr][nc] != EMPTY:
            return False
        nr += dr
        nc += dc
    return True

#해당 위치 기준으로 (시작점제외) 해당 방향으로 블럭설치 , 기본값 LINE
def set_line(r, c, dir, block = LINE) -> int:
    dr, dc = mv[dir]
    nr, nc = r + dr, c + dc
    cnt = 0
    while True:
        if nr < 0 or nr >= N or nc < 0 or nc >= N:
            break
        lst[nr][nc] = block
        cnt += 1
        nr += dr
        nc += dc
    return cnt
    
#백트래킹
def back_track(cnt, loc, line) -> None:
    global max_core
    global min_line
    # 답 갱신
    if (cnt > max_core):
        max_core = cnt
        min_line = line
    elif cnt == max_core and min_line > line:
        min_line = line
    
    # 종료 조건
    if loc == total_cnt:
        return
    
    #불필요한 탐섹 제거. 연결된 코어 + 남은 탐색 코어수 보다 현재 최대 코어보다 작은 경우 무조건 컷
    #또 같은 경우에 한해서, 현재 탐색된 라인 길이가 갱신된 최소 값보다 크거나 같으면 무시
    if total_cnt - loc + cnt < max_core or total_cnt - loc + cnt == max_core and line >= min_line:
        return
    
    for idx in range(loc, total_cnt):
        for dir in range(4):
            r, c = check_core[idx]
            if can_connet(r, c, dir):
                add_line = set_line(r, c, dir)
                back_track(cnt+1, idx+1, line+add_line)
                set_line(r, c, dir, EMPTY)
    return

for tc in range(1, T+1):
    max_core = 0
    min_line = MAX_LINE
    N = int(input())
    lst = [list(map(int,input().split())) for _ in range(N)]
    check_core.clear()
    for r in range(1, N-1):
        for c in range(1, N-1):
            if lst[r][c] :
                check_core.append((r, c,))
    total_cnt = len(check_core)
    back_track(0, 0, 0)
    
    print(f'#{tc} {min_line}')

File: SWEA/Unrated/1767. ［SW Test 샘플문제］ 프로세서 연결하기/［SW Test 샘플문제］ 프로세서 연결하기.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <algorithm>
using namespace std;
#define f(a,b,c) for (int a=b;a<c;a++)

#define CORE 1
#define LINE 2
#define EMPTY 0
#define MAXLINE 0x7f7f7f7f

int N;
int arr[12][12];
int max_core = 0, min_line = MAXLINE;
bool connected[12];
int process_loc[12][2]; //12개가 최대
int process_cnt = 0;

int dr[4] = {1, -1, 0, 0};
int dc[4] = {0, 0, 1, -1};

bool can_connect(int r, int c, int dir) {
    int nr = r + dr[dir], nc = c +dc[dir];
    while (true) {
        if (nr < 0 || nr >= N || nc < 0 || nc >= N) break;
        if (arr[nr][nc] != EMPTY) return false;
        nr += dr[dir]; nc+=dc[dir];
    }
    return true;
}

int set_line(int r, int c, int dir, int block = 2) {
    int nr = r + dr[dir], nc = c +dc[dir], cnt = 0;
    while (true) {
        if (nr < 0 || nr >= N || nc < 0 || nc >= N) break;
        ++cnt;
        arr[nr][nc] = block;
        nr += dr[dir]; nc+=dc[dir];
    }
    return cnt;
}

void back_track(int cnt, int loc, int line) { //현재 추가로 연결한 프로세서 수, 탐색시작위치, 연결된 라인 수
    //답 갱신
    if (max_core < cnt) {
        max_core = cnt;
        min_line = line;
    }
    else if (cnt == max_core &&  line < min_line) {
        min_line = line;
    }
    //탐색 종료 조건
    if (loc == process_cnt) return;

    //남은 탐색해야 할 프로세스 크기 + 이미 연결된 수가 답보다 작거나 같은데도
    //이미 라인이 답모다 크거나 같으면 탐색 의미가 없음.
    if (process_cnt - loc + cnt < max_core || process_cnt - loc + cnt == max_core && line >= min_line ) return; 
    
    //사방탐색
    for (int i = loc; i < process_cnt; ++i) {
        for (int dir = 0; dir < 4; ++dir) {
            if (can_connect(process_loc[i][0], process_loc[i][1], dir)) {
                int add_line = set_line(process_loc[i][0], process_loc[i][1], dir, LINE);
                back_track(cnt+1, i+1, line+add_line);
                set_line(process_loc[i][0], process_loc[i][1], dir, EMPTY);
            }
        }
    }
    return;
}


int main(void) {
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    // freopen("input.txt","r",stdin);
    // freopen("output.txt","w",stdout);
    int T; cin >> T;
    for (int tc = 1; tc <= T; ++tc) {
        cin >> N;
        //초기화
        max_core = 0; min_line = MAXLINE; process_cnt = 0; memset(connected, 0, sizeof(connected));
        f(r, 0 , N) {
            f(c, 0, N) {
                cin >> arr[r][c];
                if (arr[r][c]) {
                    //겉에 이미 연결되어 있는 건 무시
                    if (r == 0 || r == N-1 || c == 0 || c ==N-1) continue;
                    //연결 안된 프로세서만 등록 밑 카운팅
                    process_loc[process_cnt][0] = r;
                    process_loc[process_cnt][1] = c;
                    ++process_cnt;
                }
            }
        }
        //백트래킹
        back_track(0, 0, 0);
        cout << '#' << tc << ' ' << min_line <<'\n';
    }

    return 0;
}